### ISSUE TITLE
Validation does not produce any output in PyTorch Lightning using my UNetTestModel

### ISSUE THREAD

**Issue by lgy112112** [Lightning-AI/pytorch-lightning#19751]
### Bug description

I'm trying to validate my model using PyTorch Lightning, but no output or logs are generated during the validation process, despite setting up everything correctly.
![image](https://github.com/Lightning-AI/pytorch-lightning/assets/144128974/df1dd055-f70d-49bb-8316-cd4e2128ce58)
And this is my model part:

`class UNetTestModel(pl.LightningModule, HyperparametersMixin):
    def __init__(
        self,
        encoder_name='resnet50',
        encoder_weights='imagenet',
        in_channels=1,
        classes=14,
        loss_fn=DiceCELossWithKL(softmax=True, lambda_dice=0.85, lambda_ce=0.15, lambda_kl=2.0, to_onehot_y=True, include_background=True),
        loss_function='DiceCELossWithKL',
        learning_rate=3e-3,
        ):
        super().__init__()
        self.save_hyperparameters()
        self.model = smp.Unet(
            encoder_name=encoder_name,
            encoder_weights=encoder_weights,
            in_channels=in_channels,
            classes=classes,
        )
        self.loss_fn = loss_fn
        self.val_accuracy = torchmetrics.classification.Accuracy(task="multiclass", num_classes=14, average='macro', ignore_index=0)
        self.val_accuracy_classwise = torchmetrics.classification.Accuracy(task="multiclass", num_classes=14, average='none', ignore_index=0)
        self.Dice = torchmetrics.classification.Dice(multiclass=True, num_classes=14, average='macro', ignore_index=0)
        self.F1 = torchmetrics.classification.MulticlassF1Score(num_classes=14, average="macro", ignore_index=0)
        self.Jaccard = torchmetrics.classification.MulticlassJaccardIndex(num_classes=14, average="macro", ignore_index=0)

    def forward(self, x):
        return self.model(x)

    def training_step(self, batch, batch_idx):
        images, labels = batch
        outputs = self.forward(images)
        loss = self.loss_fn(outputs, labels.unsqueeze(1))
        self.log('train_loss', loss, on_step=True, on_epoch=False, logger=True, prog_bar=True)
        return loss

    def validation_step(self, batch, batch_idx):
        images, labels = batch
        outputs = self.forward(images)
        loss = self.loss_fn(outputs, labels.unsqueeze(1))
        accuracy = self.val_accuracy(outputs, labels)
        Dice = self.Dice(outputs, labels)
        F1 = self.F1(outputs, labels)
        Jaccard = self.Jaccard(outputs, labels)
        acc = self.val_accuracy_classwise(outputs, labels)
        self.log('val_loss', loss, on_step=True, on_epoch=False, logger=True, prog_bar=True)
        self.log('val_accuracy', accuracy, on_step=True, on_epoch=False, logger=True, prog_bar=True)
        self.log('val_F1', F1, on_step=True, on_epoch=False, logger=True, prog_bar=True)
        self.log('val_Dice', Dice, on_step=True, on_epoch=False, logger=True, prog_bar=True)
        self.log('val_Jaccard', Jaccard, on_step=True, on_epoch=False, logger=True, prog_bar=True)
        self.log('val_acc_4', acc[4], on_step=True, on_epoch=False, logger=True, prog_bar=True)
        self.log('val_acc_5', acc[5], on_step=True, on_epoch=False, logger=True, prog_bar=True)
        self.log('val_acc_10', acc[10], on_step=True, on_epoch=False, logger=True, prog_bar=True)
        self.log('val_acc_12', acc[12], on_step=True, on_epoch=False, logger=True, prog_bar=True)
        self.log('val_acc_13', acc[13], on_step=True, on_epoch=False, logger=True, prog_bar=True)
        return {"loss": loss, "accuracy": accuracy}

    def optimizer_step(self, epoch, batch_idx, optimizer, optimizer_closure, **kwargs):
        if self.trainer.global_step < 50:
            lr_scale = min(1.0, float(self.trainer.global_step + 1) / 50)
            for pg in optimizer.param_groups:
                pg["lr"] = lr_scale * self.hparams.learning_rate
        optimizer.step(closure=optimizer_closure)

    def configure_optimizers(self):
        optimizer = torch.optim.Adam(self.parameters(), lr=self.hparams.learning_rate)
        scheduler = torch.optim.lr_scheduler.CosineAnnealingLR(optimizer, T_max=5, eta_min=0.000001, last_epoch=-1)
        return {
            'optimizer': optimizer,
            'lr_scheduler': {
                'scheduler': scheduler,
                'interval': 'epoch',
                'frequency': 1,
            }
        }
`

### What version are you seeing the problem on?

v2.2

### How to reproduce the bug

```python
To view the bug, you can run the colab notebook cells (not those cells marked with Opt). You can reproduce the bug in CheckMetrics cell. It is reproduce-able in kaggle and colab thus really annoying 😡. It would be sooo much oblidged if anyone could help me with this.
https://colab.research.google.com/#fileId=https%3A//storage.googleapis.com/kaggle-colab-exported-notebooks/smu-dataset-dl-update-with-new-dataset-5df7b4b9-0565-494d-b22a-c0306ec0418e.ipynb%3FX-Goog-Algorithm%3DGOOG4-RSA-SHA256%26X-Goog-Credential%3Dgcp-kaggle-com%2540kaggle-161607.iam.gserviceaccount.com/20240410/auto/storage/goog4_request%26X-Goog-Date%3D20240410T014637Z%26X-Goog-Expires%3D259200%26X-Goog-SignedHeaders%3Dhost%26X-Goog-Signature%3D0e9d4ad91ecc6e2fae0a51622b97399160be483c4737424d1584ebdcba2a80b870f32feacc256675774b85db2c72329819040ffa6c923e20835b331d995cfea132418460df7cfba6d261e6e3381354d8ca92188ddba7e502fa71fee33c63ed5d5246df0964c3766b7a26c92b559e3e359f4bc4e78b78edf3114d0d52ab54244f7c28b560f6a31a14389b27cb86837fcfb0579c6784958ab181af41a2a915a57eaa6e0e80bc9acc55bca97cbc0311caa0e870004659e568e2acae6de0af29ff8f08bbc9ebea6118b8b9d48aea9d20593a1e3516763105e0c296679a649968501b481f722936008f893bacf0856e288c202e3124902da7cdf635d174169c05b27a&scrollTo=HbUeiVEzr21d&line=4&uniqifier=1
```

### Error messages and logs

![image](https://github.com/Lightning-AI/pytorch-lightning/assets/144128974/e0a950cf-2812-4e35-8fcf-64d49e2ee982)

### Environment

basic Colab env with pip-qqq-accessible lightning
